contributors/design-proposals/hugepages.md

```diff
@@ -0,0 +1,281 @@
+limitations.  In particular, they are limited to 2MB page sizes.  
+
+Managing memory is hard, and unfortunately, there is no one-size fits all
+solution for all applications.  THP might lead to performance degradation on
+nodes with high memory utilization or fragmentation due to the defragmenting
+efforts of THP, which can lock memory pages.  For this reason, some applications
+may be designed to (or recommend) usage of pre-allocated huge pages instead of
+THP. 
+
+## Scope
+
+This proposal only includes pre-allocated huge pages configured on the node by
+the administrator at boot time or by manual dynamic allocation.  It does not
+discuss how the cluster could dynamically attempt to allocate huge pages in an
+attempt to find a fit for a pod pending scheduling.  It is anticipated that
+operators may use a variety of strategies to allocate huge pages, but we
+do not anticipate the kubelet itself doing the allocation.  Allocation of huge
+pages ideally happens soon after boot time.
+
+A simple solution via a `DaemonSet` is presented here:
+https://github.com/derekwaynecarr/hugepages/tree/master/allocator
+
+This proposal defers issues relating to NUMA.
+
+## Use Cases
+
+The class of applications that benefit from huge pages typically have
+- A large memory working set
+- A sensitivity to memory access latency
+
+Example applications include:
+- Java applications can back the heap with huge pages using the
+  `-XX:+UseLargePages` and `-XX:LagePageSizeInBytes` options.
+- database management systems (MySQL, PostgreSQL, MongoDB, etc.)
+- packet processing systems (DPDK)
+
+Applications can generally use huge pages by calling
+- `mmap()` with `MAP_ANONYMOUS | MAP_HUGETLB` and use it as anonymous memory
+- `mmap()` a file backed by `hugetlbfs`
+- `shmget()` with `SHM_HUGETLB` and use it as a shared memory segment (see Known
+  Issues).
+
+1. A pod can use huge pages with any of the prior described methods.
+1. A pod can request huge pages.
+1. A scheduler can bind pods to nodes that have available huge pages.
+1. A quota may limit usage of huge pages.
+1. A limit range may constrain min and max huge page requests.
+
+## Feature Gate
+
+The proposal introduces huge pages as an Alpha feature.
+
+It must be enabled via the `--feature-gates=HugePages=true` flag on pertient
+components pending graduation to Beta.
+
+## Node Specfication
+
+Huge pages cannot be overcommitted on a node.
+
+A system may support multiple huge page sizes.  It is assumed that most
+nodes will be configured to primarily use the default huge page size as
+returned via `grep Hugepagesize /proc/meminfo`.  This defaults to 2MB on
+most Linux systems unless overriden by `default_hugepagesz=1g` in kernel
+boot parameters.  This design does not limit the ability to use multiple
+huge page sizes at the same time, but we expect this to not be the normal
+setup.  For each supported huge page size, the node will advertise a
```

Comment:
nit: not sure if we need this sentence (starting with "This design..")  i found it confusing i.e. you can, but maybe you shouldn't because... reasons?